### PR TITLE
Implementation of DoFTools::locally_relevant_dofs_per_subdomain

### PR DESCRIPTION
**Pull request by jppelteret**
To assist in the situations where grid partitioning is done with METIS (i.e. standard `Triangulation`s via `GridTools::partition_triangulation` or `parallel::shared::Triangulation`), a new function `DoFTools::locally_relevant_dofs_per_subdomain` has been introduced.

Since it was missing, I also added Denis Davydov's authorship for `DofTools::locally_owned_dofs_per_subdomain`

### CODE REVIEW

File: source/dofs/dof_tools.cc [hunk 1]
```diff
@@ -1124,6 +1125,60 @@ namespace DoFTools
...
+                   ExcMessage("The subdomain ID of the halo cell should not match that of the vector entry."));
+
+            local_dof_indices.resize(cell->get_fe().dofs_per_cell);
+            cell->get_dof_indices(local_dof_indices);
+
+            for (std::vector<types::global_dof_index>::iterator it=local_dof_indices.begin();
+                 it!=local_dof_indices.end();
+                 ++it)
+              if (!dof_set[subdomain_id].is_element(*it))
+                subdomain_halo_global_dof_indices.insert(*it);
```
Comment (bangerth) [inline source/dofs/dof_tools.cc]: couldn't you just save yourself the `if` statement and simply insert all `local_dof_indices` here? Duplicates will be filtered out later on when you merge with `dof_set`.

**Comment by bangerth** [comment]
Looks mostly good. Since the function doesn't work with `p::d::Triangulation`, can you add an assertion that verifies this? Also, having a test would be good.

File: source/dofs/dof_tools.cc (hunk 1 above)
Comment (jppelteret) [inline source/dofs/dof_tools.cc]: Sure, no problem!

**Comment by jppelteret** [comment]
Cool. I've discussed with Denis how to devise a test for this. I'll implement it (hopefully) soon.

**Comment by bangerth** [comment]
It doesn't have to be anything difficult. Just clone one of the existing tests in the shared tria directory where we already build a DoFHandler in parallel, then call this function and output the result.

**Comment by bangerth** [comment]
J-P: Does the test actually work on your machine? It seems to provide a different answer on the "official" tester:

http://cdash.kyomu.43-1.org/testDetails.php?test=23274485&build=13117

**Comment by davydden** [comment]
i think it goes back to un-repeatable METIS (used inside parallel::shared::Tria) and the discussion here https://github.com/dealii/dealii/pull/1412 . I think you can add the "official" tester output as another .output file to make it pass.

**Comment by bangerth** [comment]
Ah, good point. Will take care of it.

**Comment by bangerth** [comment]
See #1698.

**Pull request merged**